addon/devtoolsPanel.js

```diff
@@ -1,47 +1,232 @@
+        row.appendChild(td);
+    }
+
+    function removeLabeled(event) {
+        event.preventDefault();
+        backgroundPort.postMessage({
+            type: 'label',
+            tabId: browser.devtools.inspectedWindow.tabId,
+            selector: event.target.dataset.path,
+            label: '',
+        });
+    }
+
+    function addClearLabelCell(row, label) {
+        const td = document.createElement('td');
+        td.classList.add('action');
+        const a = document.createElement('a');
+        a.href = '#';
+        a.classList.add('action-button');
+        a.dataset.path = label.path;
+        a.dataset.label = label.label;
+        a.setAttribute('title', label.inspected ? 'Clear Label' : 'Remove Label');
+        a.addEventListener('click', removeLabeled);
+        a.appendChild(document.createTextNode(label.inspected ? '⦸' : '✖'));
+        td.appendChild(a);
+        row.appendChild(td);
+    }
+
+    const table = document.createElement('table');
+    const headerRow = document.createElement('tr');
+    addTextCell(headerRow, 'Selector', true).classList.add('path');
+    addTextCell(headerRow, 'Element', true).classList.add('preview');
+    addTextCell(headerRow, 'Label', true).classList.add('label');
+    addTextCell(headerRow, '').classList.add('action');
+    table.appendChild(headerRow);
+
+    for (const label of labeled) {
+        const row = document.createElement('tr');
+        row.classList.add('hover');
+
+        addPathCell(row, label);
+        addTextCell(row, label.preview).classList.add('preview');
+        addLabelCell(row, label);
+        addClearLabelCell(row, label);
+
+        table.appendChild(row);
+    }
+
+    const container = document.getElementById('labels');
+    emptyElement(container);
+    container.appendChild(table);
 }
-document.getElementById('saveButton').addEventListener('click', freezePage);
-
-/**
- * Update the GUI to reflect the currently inspected page the first time the
- * panel loads.
- *
- * This runs once per Fathom dev-panel per inspected page. (When you navigate
- * to a new page, the Console pane comes forward, so this re-runs when the
- * Fathom pane is brought forward again. It does not run twice when you switch
- * away from and then back to the Fathom dev panel. Contents of the dev panel
- * are preserved across that interaction.)
- */
-async function initPanel() {
-    // When we load a new page with existing annotations:
-    //browser.devtools.inspectedWindow.eval(`document.querySelectorAll("[data-fathom]")[0].id`).then(updateUi);
+
+// Update the UI to show either the table of labeled elements, or
+// a message informing the user to use the inspection panel.
+function updateLabeled() {
+    getLabeled()
+        .then((labeled) => {
+            updateLabeledTable(labeled);
+
+            if (labeled.length === 0) {
+                // nothing inspected or already labeled
+                document.getElementById('no-selection').classList.remove('hidden');
```

Comment:
Okay, that's a lot of `document.getElementById(…).classList`! Time for some factoring out yet? Maybe a 3rd-party lib? :-) For a later PR, of course.